### PR TITLE
[WIP] Add Unix implementation for System.Drawing.Common

### PR DESCRIPTION
**Pull request by mellinoe**
This is a port of mono's System.Drawing implementation, for use with .NET Core. The goal is to share this source code directly with mono, and build for all configurations from the same codebase.

In the present version, the code files have stayed in the same folder structure that they had in mono. This made it easier for me to make sure that mono's build continued to work using the same files. Ultimately, we should condense the folders down to match other corefx projects, and condense implementations as much as possible, as well.

Open questions:

* What should we do about `[MonoTODO]` attributes? 
* ~~What should we do about the `Locale` usage?~~ Removed.
* What icons should be used in the `SystemIcons` class? I've taken the "Error" icon from mono and returned it for all system icons currently.

@marek-safar @qmfrederik @hughbe

### CODE REVIEW

**Comment by danmosemsft** [comment]
Presumably some of the code can be coalesced with Windows over time. For example KnownColors/KnownColorTable is just C# tables. How do you see that working -- we would share one copy out of the Mono subdirectory (presumably the Windows one to minimize risk to the most widely used version)?

**Comment by mellinoe** [comment]
Eventually, I envision that we will re-use large chunks of the code, and all of it will live in the "regular folder structure", with OS-specific files, and partial classes, etc. used where appropriate. Like you mentioned, there's a lot of code that can be fully shared, like enums, exceptions, static data, etc. @qmfrederik started some of that with this PR to mono:  https://github.com/mono/mono/pull/4980. That was for mono's build configuration, but we can essentially start doing the same for the Windows/Unix configurations for .NET Core, and then flow the consolidation back into mono.

**Comment by akoeplinger** [comment]
> What should we do about the Locale usage?

We never localized System.Drawing (at least as far as I'm aware of) so we could just remove those usages from Mono too.

**Comment by mellinoe** [comment]
@akoeplinger Great, that simplifies things on that front.

**Comment by akoeplinger** [comment]
@mellinoe should I do the update in Mono or would that mess with your WIP?

**Comment by mellinoe** [comment]
@akoeplinger I believe there are only a small amount of `Locale` usages in the library (~50), so the diff won't be hard to reconcile. But, I have made a lot of code style changes in this version already, so it may be best to make the changes here.

**Comment by akoeplinger** [comment]
@mellinoe ok, whatever is easier for you :)

**Comment by mellinoe** [comment]
Removed usages of Locale: https://github.com/dotnet/corefx/pull/20887/commits/873c22eeae2f124280fa45c92f610938469683ff

File: src/System.Drawing.Common/src/mono/System.Drawing/Icon.cs [hunk 1]
```diff
@@ -0,0 +1,963 @@
...
+using System.Drawing.Imaging;
+using System.IO;
+using System.Reflection;
+using System.Runtime.Serialization;
+using System.Runtime.InteropServices;
+
+namespace System.Drawing
+{
+    [Serializable]
+#if !NETCORE
```
Comment (hughbe) [inline src/System.Drawing.Common/src/mono/System.Drawing/Icon.cs]: Do we still need this and the one below?

Comment (mellinoe) [inline src/System.Drawing.Common/src/mono/System.Drawing/Icon.cs]: Yes, we don't want these attributes in the .NET Core version.

**Comment by hughbe** [comment]
Can you change the active issues to be unix specific?

**Comment by mellinoe** [comment]
>Can you change the active issues to be unix specific?

Oops. Yes, I will do that 😄

**Comment by mellinoe** [comment]
I've rebased and made a couple more minor changes to this. I'm planning on merging this once the CI is green, and will continue tracking the related follow-up issues.

**Comment by mellinoe** [comment]
Actually, we will need to install `libgdiplus` on all of our CI machines, so this will need to wait to be merged. I will update the PR once that dependency has been installed everywhere.

File: src/System.Drawing.Common/src/mono/System.Drawing/Icon.cs (hunk 1 above)
Comment (ViktorHofer) [inline src/System.Drawing.Common/src/mono/System.Drawing/Icon.cs]: Can you please make sure to remove all the serializable attributes here and in System.Drawing.Common?

Comment (qmfrederik) [inline src/System.Drawing.Common/src/mono/System.Drawing/Icon.cs]: If the plan is that Mono will consume these sources "as is", you'll probably need to put them behind a feature flag (as opposed to removing them altogether)

**Comment by qmfrederik** [comment]
@mellinoe do you plan to backport the changes done in the Mono repo to this branch before merging this, or merge first and then apply them in a separate PR? The main advantage if the work done over at Mono is that a lot of files can be shared accross Unix & Windows.

Also let me know if you plan to bulk import Mono's unit tests, happy to help. I'm a bit concerned some of the cleanup done here may break Mono stuff :)

**Comment by mellinoe** [comment]
@qmfrederik Yes, I will sync up with the code that is in mono before merging this.

> Also let me know if you plan to bulk import Mono's unit tests, happy to help. I'm a bit concerned some of the cleanup done here may break Mono stuff :)

We have several people working on moving mono's tests over piece-by-piece. I'm not sure it makes sense to do a big "code dump" to get them over here, because a lot of test cases require modifications to pass cleanly anyways. And that's beyond the obvious Nunit-Xunit conversion and other style changes needed. I think it makes sense to continue porting tests in whatever-size chunks folks are comfortable working with, and eventually retire the old mono tests, as long as we believe the coverage is strictly greater here in corefx. I think we have made pretty decent progress so far with the current set of tests we've ported.

**Comment by hughbe** [comment]
I'm happy with the way @norek and I are adding tests now. Not only are we porting mono tests and making them cleaner, but we're adding missing coverage along the way, which a full dump would disincentivise.

**Comment by karelz** [comment]
@mellinoe what is the next step for this PR? Is it blocked? Or just on back burner due to higher priority stuff?

**Comment by mellinoe** [comment]
@karelz A little bit of both. I put it off to let other changes go through first, mainly some work to replace some parts of mono's implementation with ours. But I haven't had time to circle back and see if I can rebase this yet.

**Comment by karelz** [comment]
If you think you won't have time in next week or two, let's close it, until there is right time for it ...

**Comment by mellinoe** [comment]
I'll re-create this when I have more time. Probably next week.

**Pull request closed without merging**